Support for secret store reference in Dapr components

SoTrx opened the pull request:
# Description
Hey,

This PR adds support for building Dapr components with an optional reference to a Dapr secret store.

The new property is named _secretStoreComponentName_. This name is totally arbitrary but was chosen to try to be as clear as possible.

A secret reference can be created using a _secretKeyRef_ object, following the same [naming convention as Dapr](https://docs.dapr.io/operations/components/component-secrets/).

Example : 
```yaml
resource stateStore 'Applications.Dapr/stateStores@2023-10-01-preview' = {
  name: 'statestore'
  properties: {
    environment: environment
    application: application
    resourceProvisioning: 'manual'
    type: 'state.redis'
    version: 'v1'
    metadata: {
      redisHost: {
        secretKeyRef: {
            name: 'redisHost'
            key: 'redisHost'
        }

      }
      redisPassword: ''
    }
    secretStoreComponentName: daprSecretStore.name
  }
}
```

## Type of change

- This pull request adds or changes features of Radius and has an approved issue (issue link required).

Fixes: #7704

rynowak commented:
Wow thanks for sending this. This is a big improvement. I'll take a look at this tomorrow. 

Normally we might ask for an design document for an API change, but I think we can skip that in this case:

- It's a very small change
- It has an approved issue that describes the API change.
- It's aligned with the naming and design of the existing Dapr API - which we're trying to match.

@radius-project/maintainers-radius holler if you have any concerns.

ytimocin commented on pkg/daprrp/processors/statestores/processor_test.go:
```diff
@@ -199,6 +200,111 @@ func Test_Process(t *testing.T) {
+						},
+					},
+				},
+				"auth": map[string]any{
+					"secretStore": secretStoreComponentName,
+				},
```
Is this `auth` generated by **Dapr** since we create a **Dapr** object with a property called **secretStoreComponentName**?

SoTrx replied:
In this case, the auth.secretstore property is added [here](https://github.com/radius-project/radius/pull/7823/files/4b23f1c2bb1e783d06e3d11a34c2296dbb512e85#diff-6ae8441cbc7aeb3a6c3a1d2dc2e278e964ddf0f99649c07832c7bf133f436e7dR100-R107) if the Applications.Dapr/* contains the `secretStoreComponentName` property.
The auth.secretstore property comes from the [Dapr component spec](https://github.com/dapr/dapr/blob/master/charts/dapr/crds/components.yaml#L26)

ytimocin commented on pkg/portableresources/renderers/dapr/generic.go:
```diff
@@ -56,9 +57,24 @@ func ConstructDaprGeneric(daprGeneric DaprGeneric, namespace string, componentNa
+
+		switch value := v.(type) {
+
+		// Nested object, has to be a secret reference
+		case map[string]any:
+			if secretKeyRef, ok := value["secretKeyRef"]; ok {
+				yamlItem["secretKeyRef"] = secretKeyRef
+			} else {
+				return unstructured.Unstructured{}, v1.NewClientErrInvalidRequest(fmt.Sprintf("Invalid metadata value for key %s in Dapr component %s", k, componentName))
+			}
```
Could there be a case where it is not a secret reference even if it is a nested object? Do we need to do an additional check for this case to make sure that it is a secret reference?

rynowak replied:
You can see the schema of `Component` here: https://github.com/dapr/dapr/blob/master/charts/dapr/crds/components.yaml#L48

Note that this is the **output** we're generating. 

The reason why I'm mention it is that it shows what the schema is in Dapr that we're trying roughly match. The allowed keys are:

- `value`
- `secretKeyRef`
- `envRef`

`envRef` isn't relevant for us - it's a dev-time feature.

ytimocin commented on typespec/Applications.Dapr/secretStores.tsp:
```diff
@@ -45,10 +45,15 @@ model DaprSecretStoreResource
   name: ResourceNameString;
 }
 
+// This is to prevent a secret store from referencing another secret store
+// This would be valid according to Dapr component specs, but may confuse users
+@withoutOmittedProperties("secretStoreComponentName")
+model NonRedundantDaprResourceProperties is DaprResourceProperties;
```
Could there be a case where users can require authentication to a Secret Store?

rynowak replied:
I think we're assuming the Dapr secret store has already been created.

ytimocin reviewed (commented):
This is a great addition. Thanks!

Should we add a functional test or add this to [an existing functional test](https://github.com/radius-project/radius/blob/main/test/functional-portable/daprrp/noncloud/resources/testdata/daprrp-resources-statestore-manual.bicep) so that we can test it out?

rynowak commented:
Hi @SoTrx - thanks for sending this PR. Looks like it works overall.

One question based on your example:

```bicep
resource stateStore 'Applications.Dapr/stateStores@2023-10-01-preview' = {
  name: 'statestore'
  properties: {
    environment: environment
    application: application
    resourceProvisioning: 'manual'
    type: 'state.redis'
    version: 'v1'
    metadata: {
      redisHost: {
        secretKeyRef: {
            name: 'redisHost'
            key: 'redisHost'
        }

      }
      redisPassword: ''
    }
    secretStoreComponentName: daprSecretStore.name
  }
}
```

It looks like the type of `metadata` after your change is `map from string to (string or object)`. This isn't ideal because we try to avoid designs where a field has an ambiguous type. Ideally this would be `map from string to object` in all cases. I think we probably made a mistake in the original design 😢 

If I compare to the Dapr schema of `Component` I'd describe this as `list of key-value-pairs where the key is a string and the value is an object`.

Do you think it makes sense for us to change the design?

Example:

```bicep
resource stateStore 'Applications.Dapr/stateStores@2023-10-01-preview' = {
  name: 'statestore'
  properties: {
    environment: environment
    application: application
    resourceProvisioning: 'manual'
    type: 'state.redis'
    version: 'v1'
    metadata: {
      redisHost: {
        secretKeyRef: {
            name: 'redisHost'
            key: 'redisHost'
        }
      }
      redisPassword: {
        value: ''
      }
    }
    secretStoreComponentName: daprSecretStore.name
  }
}
```

This would be a breaking change, and we'd need to update our existing code, but it's probably a better design. What do you think?

SoTrx commented:
Thank you for your reviews

@rynowak 
I believe your proposed change might be the best approach, as it would provide the most extensibility.

Going the other way and using strings only (with a magic prefix to reference secrets) might also seem tempting since it would be non-breaking. But this is a  worse design and may lead to parsing/escaping issues or conflict with [special values](https://docs.dapr.io/reference/resource-specs/component-schema/#templated-metadata-values) or even a company's own oddly defined [pluggable components](https://docs.dapr.io/developing-applications/develop-components/pluggable-components/pluggable-components-overview/).

```bicep
resource stateStore 'Applications.Dapr/stateStores@2023-10-01-preview' = {
  name: 'statestore'
  properties: {
    environment: environment
    application: application
    resourceProvisioning: 'manual'
    type: 'state.redis'
    version: 'v1'
    metadata: {
      redisHost: 'secretRef:://key,value'
      redisPassword: ''
    }
    secretStoreComponentName: daprSecretStore.name
  }
}
```

Another approach could be treating the secret as its own resource. However, aside from the added complexity, this might encourage users to attempt using secrets from multiple secret stores within a single component, which, as far as I'm aware, Dapr does not support.

So using only objects may be the best solution in my opinion.

------------

On a related note, if the goal is to match Dapr component schema, should we also change the `secretStoreComponentName` property to `auth.secretstore` to match what Dapr [uses]((https://docs.dapr.io/operations/components/component-secrets/) ?

Something like :

```bicep
resource stateStore 'Applications.Dapr/stateStores@2023-10-01-preview' = {
  name: 'statestore'
  properties: {
    environment: environment
    application: application
    resourceProvisioning: 'manual'
    type: 'state.redis'
    version: 'v1'
    metadata: {
      redisHost: {
        secretKeyRef: {
            name: 'redisHost'
            key: 'redisHost'
        }
      }
      redisPassword: {
        value: ''
      }
    }
    auth: {
      secretStore: daprSecretStore.name
    }
  }
}
```

---------

I'll make some functional tests with the curent implementation first as @ytimocin suggested before going forward with the changes we decide on.

SoTrx committed:
feat: parse all Dapr components metadata as objects

BREAKING CHANGE: All Dapr components metadata are now parsed as objects instead of their previous format. All string-based values will break

Signed-off-by: SoTrx <11771975+SoTrx@users.noreply.github.com>

SoTrx commented:
The current implementation follow this schema :
```bicep 
resource stateStore 'Applications.Dapr/stateStores@2023-10-01-preview' = {
  name: 'statestore'
  properties: {
    environment: environment
    application: application
    resourceProvisioning: 'manual'
    type: 'state.redis'
    version: 'v1'
    metadata: {
      redisHost: {
        secretKeyRef: {
            name: 'redisHost'
            key: 'redisHost'
        }
      }
      redisPassword: {
        value: ''
      }
    }
    secretStoreComponentName: daprSecretStore.name
  }
}
```
Should I change secretStoreComponentName to auth.secretStore as well? 

Let me know if you need any further adjustments

rynowak commented:
> Should I change secretStoreComponentName to auth.secretStore as well?
> 
> Let me know if you need any further adjustments

I'm on board with this 👍. I think it's good if we align to the Dapr component schema. Then it will be easy to update Radius as Dapr continues to add features.

Taking a look now! Thanks for working through this.

rynowak commented on test/testrecipes/modules/redis-selfhost.bicep:
```diff
@@ -44,12 +52,11 @@ resource redis 'apps/Deployment@v1' = {
+                      ['redis-cli'],
+                      password != '' ? ['-a', password] : [],
+                      ['-h', 'localhost', 'MONITOR']
+                  )
```
nice 👍

rynowak reviewed (changes requested):
This is looking awesome overall. Can you make the schema changes?

SoTrx committed:
Properly typing Dapr components Metadata
Signed-off-by: SoTrx <11771975+SoTrx@users.noreply.github.com>

SoTrx committed:
feat(backend): add types to all Dapr components metadata fields in daprrp
Signed-off-by: SoTrx <11771975+SoTrx@users.noreply.github.com>

SoTrx commented on typespec/Applications.Dapr/common.tsp:
```diff
@@ -33,4 +33,7 @@ model DaprResourceProperties {
 
   @doc("Dapr component version")
   version?: string;
+
```
Sure! I'll also add types to the Metadata field in the Dapr components data model as well, if that’s okay with you. This might help prevent some "casting to any" errors.

SoTrx committed:
refactor: replace secretStoreComponentName with auth.secretStore in all Dapr components

BREAKING CHANGE: The property `secretStoreComponentName` has been replaced with `auth.secretStore`. This change requires updates to any references to the old property.
Signed-off-by: SoTrx <11771975+SoTrx@users.noreply.github.com>

SoTrx committed:
fix: Lint TypeSpec files

Signed-off-by: SoTrx <11771975+SoTrx@users.noreply.github.com>

SoTrx committed:
style: No running go fmt for generated files zz_generated_models.go zz_generated_models_serde.go

Signed-off-by: SoTrx <11771975+SoTrx@users.noreply.github.com>

rynowak reviewed (approved):
Wow this looks great! Thanks @SoTrx

kachawla commented:
@SoTrx Thank you for contributing. We have updated the [samples](https://github.com/radius-project/samples/pull/1757/files) to reflect these changes. Could you please update the [docs](https://github.com/radius-project/docs/tree/edge) corresponding to this feature? For example, [here](https://github.com/radius-project/docs/blob/edge/docs/content/guides/author-apps/dapr/overview/snippets/statestore.bicep#L37).
